Nimble Rich Text Editor and Viewer Spec Document (#1289)
ni/nimble: packages/nimble-components/src/rich-text-editor/specs/README.md

    of the comments feature. Once the visual design for these components is complete, we will then be implementing those specific changes within the defined
    scope of development. However, we will still make use of existing nimble components such as `nimble-button` and `nimble-text-area` to maintain a
    consistent design for the initial release.

### Prior Art/Examples

Sample Mockup Screenshots

_`nimble-rich-text-editor`_

![Editor](./spec-images/editor-sample.png)

_`nimble-rich-text-viewer`_

![Viewer](./spec-images/viewer-sample.png)

## Design

### `nimble-rich-text-editor`

The `nimble-rich-text-editor` will be divided into two sections namely an `editor` section and a `footer` section.

1. `editor` section is the actual text area to add or update rich text content.
2. `footer` section consists of `nimble-button` to control each text formatting functionalities like bold, italic, etc, and a
   `footer-actions` slot element which is typically used to add action buttons to the right bottom of the component.

Example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>

>     <div slot="footer-actions">

Reviewer: Is it expected that a client would provide a `div` containing multiple buttons like this, or would they do something more like:
```
<nimble-rich-text-editor>
    <nimble-button slot="footer-actions>Cancel</nimble-button>
    <nimble-button slot="footer-actions>Add Comment><nimble-button>
</nimble-rich-text-editor>
```

In the example you provide, it would be the responsibility of a client to space the buttons appropriately. However, in the example I provided, the nimble component could add the appropriate spacing between each component placed in the `footer-actions` slot.

Author: I have updated the spec as per the example you have showed. The standard spacing between the slotted elements will be handled by the component in `footer-actions` sections and other styling will be handled by the client.

```

### API

_Props/Attrs_

-   `markdownValue` - is an accessor used to get and set the markdown value.
    -   `getter` - this will serialize the content by extracting the Node from the editor and convert it into a markdown string using
        [prosemirror-markdown serializer](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/to_markdown.ts#L30).
    -   `setter` - this will parse the markdown string into a Node and load it back into the editor using
        [prosemirror-markdown parser](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/from_markdown.ts#L199).

_Methods_

-   none

_Events_

-   `input` - event emitted when there is a change in the editor. This can be achieved through Tiptap's [update event](https://tiptap.dev/api/events#update).
    Below are a few scenarios to understand when this event will fire:

    1. This event will fired for every input in the content of the editor, including text inputs, text formatting changes, and text removals.
    2. This event will not fire when there are no changes made to the content of the editor. For example, all mouse events, selecting the texts, state
       changes, etc,

>     The emitted event data is in the format of a `CustomEvent` and consists of a `detail` object with an `isEmpty` boolean flag indicating whether the editor

Author: Thanks for bringing up this use case. I have updated `isEmpty` as a read-only property as suggested and removed it from the event data.

    is empty or not.

_CSS Classes and CSS Custom Properties that affect the component_

> -   The minimum and maximum height of the `editor section` will be fixed value. If the content exceeds the maximum height, we will show the vertical scrollbar to

Author: I have updated it accordingly in the spec to have the minimum height as per the suggestion from Jesse in the other comment and the minimum width based on to accommodate all buttons being visible in the footer. 

The maximum width and height will not be fixed in the component and will be configurable by the client.

> -   The minimum and maximum height of the `editor section` will be fixed value. If the content exceeds the maximum height, we will show the vertical scrollbar to

Author: Thanks for the detailed information. I have updated the spec as per the suggestion. We will get started with this approach for the initial release and as said as per the visual design from Rick, we will incorporate the changes during the implementation or in the subsequent releases.

> -   The minimum and maximum height of the `editor section` will be fixed value. If the content exceeds the maximum height, we will show the vertical scrollbar to

Reviewer 2: Looking good, thanks for updating. A couple last minor points:

>The client will determine the default and maximum height and width of the component.

Nimble should provide a default height and width for the case where the element is placed on the page with no size styling. This can just be a reasonable fixed pixel size, as long as the client can override it.

Also, a possible nice-to-have feature for the editor component would be to use the [built-in resize attribute](https://developer.mozilla.org/en-US/docs/Web/CSS/resize) to allow the user to dynamically grow the height of the text input box (like GitHub allows). I bet it would be trivial to implement, but it's just an idea so no need to capture it if you're worried about scope.

![image](https://github.com/ni/nimble/assets/10500124/e4c135a7-94ed-4efa-afce-e005d6fa6a3c)

> -   The minimum and maximum height of the `editor section` will be fixed value. If the content exceeds the maximum height, we will show the vertical scrollbar to

Reviewer 3: Thanks for suggesting this @jattasNI. This new feature looks great. And from your reference, as you mentioned it seems easy to integrate. Can we please consider this after completing all the required features? As I am not sure of the development effort to write test scripts for this feature, it would be easy for us to analyze this once after making good progress on the development.

> -   The minimum and maximum height of the `editor section` will be fixed value. If the content exceeds the maximum height, we will show the vertical scrollbar to

Reviewer 2: Definitely, that makes sense.

    view the content of the editor.
-   The minimum and maximum width of the `editor section` will be fixed value considering the mobile and desktop view of the component respectively. However, if the
    screen width is reduced
    below the specified minimum width, the layout may not handle or adapt to the content.
-   The `footer section` will be a flexbox container and have the same width as the editor section.
-   The `formatting option toolbar` in the footer section will be enclosed within a flexbox container, enabling the buttons to wrap and occupy roughly seventy

>     percent of the entire footer. If the buttons exceed the seventy percent limit, they will be positioned below in the same container, aligning from left

Author: This will be applicable only if we set the maximum height in the component and not in the client. Wrapping the buttons and adding them to the other row will increase the height of the component. So, we thought of having a minimum width for the entire component based on the buttons available in the footer (formatting options + slot elements). For the initial release, we will consider having the four buttons space for the formatting options and the remaining space for the slot as mentioned in the spec.

       there are additional elements beyond this limit, they will be positioned below within the same footer-actions container.

_Note_: The positioning of these slot elements in the mobile view of the component has not yet been confirmed.

_Host Classes_

-   none

_Slotted Content/Slotted Classes_

-   none

_CSS Parts_

-   none

### `nimble-rich-text-viewer`

The `nimble-rich-text-viewer` is used for viewing rich text content when a markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### API

_Props/Attrs_

-   `markdownValue` - is an accessor used to get and set the markdown value.
    -   `getter` - this will merely return the markdown string that is set to the component.
    -   `setter` - this will parse the markdown string into a Node using
        [prosemirror-markdown parser](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/from_markdown.ts#L199) and convert to an HTML to
        render it in the component section.

_Methods_

-   none

_Events_

-   none

_CSS Classes and CSS Custom Properties that affect the component_

> -   The height of the component will always grow to fit the content.

Reviewer 2: It would be good to clarify: clients can still set the width? And this setting could affect the height (a smaller width will cause the content to reflow and take up more height)?

Author: Yes, it will increase the height if the width is small. I have updated the same in the spec.

-   The minimum and maximum width of the component will be the same as the `nimble-rich-text-editor`, considering the mobile and desktop views respectively.

### Anatomy

_Slot Names_

-   none

_Host Classes_

-   none

_Slotted Content/Slotted Classes_

-   none

_CSS Parts_

-   none

### Angular integration

An Angular directive will be created for both components. The components will not have form association, so a `ControlValueAccessor` will not be created.

### Blazor integration

A Blazor wrapper is initially out of scope for these components.

### Visual Appearance

> NA

Reviewer 2: ```suggestion
Visual design has been requested is currently still pending.
```

[marks](https://tiptap.dev/api/marks) and [nodes](https://tiptap.dev/api/nodes).

-   The term `Marks` in the Tiptap are formatting styles applied to specific portions of text, such as bold, italic, or underlined.
-   `Nodes` in the Tiptap editor are the building blocks of the document structure, representing different types of content elements like paragraphs,
    headings, lists, images, and more.

The rich text content entered in the editor is converted to markdown output using
[prosemirror-markdown](https://github.com/ProseMirror/prosemirror-markdown) serializer. Here is the reference for the supported formatting schema in the
markdown based on [CommonMark](http://commonmark.org/) flavor:

-   Bold - `**Bold**`
-   Italics - `*Italics*`
-   Numbered list - `1. Numbered list`
-   Bulleted list - `* Bulleted list`

The `nimble-rich-text-viewer` will be responsible for converting the input markdown string to HTML Fragments with the help of
`prosemirror-markdown` parser, which is then converted to HTML string and rendered into the component to view all rich text content.

> ### Prototype

Reviewer 2: It'll be good to have a discussion sometime about the process for converting this prototype into production code. I don't think we want one big PR for the entire feature at once. Rather it would be good to start small with a vertical slice and incrementally add functionality. This can help ensure better code quality and test coverage.

No spec updates are needed for this, just want to put it on your radar as a future discussion point.

Author: Sure, we will plan to discuss this once we start with the implementation and make sure to have a more granular PR for the entire component development.

[nimble-toggle-button](https://nimble.ni.dev/storybook/?path=/docs/toggle-button--docs) to determine whether it's in an active or inactive state.
Below is the sample screenshot which shows the active state of the `nimble-toggle-button` when the cursor is focused on the `Bold` and `Numbered` List`
text in the editor.

![Bold and Numbered Button State](./spec-images/button-state.png)

### Accessibility

[Tiptap Accessibility](https://tiptap.dev/guide/accessibility)

[Toolbar Accessibility](https://www.w3.org/WAI/ARIA/apg/patterns/toolbar/)

[Button Accessibility](https://www.w3.org/WAI/ARIA/apg/patterns/button/)

_Focus_

-   Focus state of the editor will be the same as the `nimble-text-area`.
-   Focus state of the buttons will be the same as the `nimble-button`.

_Keyboard accessibility and shortcuts for text formatting_

All the basic keyboard accessibility for the tiptap rich text editor is mentioned in their
[keyboard shortcut](https://tiptap.dev/api/keyboard-shortcuts#predefined-keyboard-shortcuts) page. To facilitate easy access, listed down
the supported text formatting keyboard shortcuts in the tiptap editor.

> | Key                     | Behavior                                                 |

Reviewer 2: Double checking: does TipTap automatically support Command instead of Ctrl for the shortcuts on macOS?

Author: Yes, it does support `Cmd` in macOS, and here is the official page of Tiptap for [keyboard shortcuts](https://tiptap.dev/api/keyboard-shortcuts#predefined-keyboard-shortcuts). I have added this point as well in the spec.

### Security

-   Prose mirror uses markdown-it for converting markdown to HTML and HTML to markdown. We will follow the
    [security guidelines of markdown-it](https://github.com/markdown-it/markdown-it/blob/master/docs/security.md#security) to turn
    off HTML at source as given in the [API docs](https://markdown-it.github.io/markdown-it/#MarkdownIt.new). Prosemirror-markdown follows the same as
    shown in this
    [specific line of code](https://github.com/ProseMirror/prosemirror-markdown/blob/26e58302399b7d9a9b3bc8fc3bf5807627ca29e5/src/from_markdown.ts#L245).
-   For additional safety, we may use [sanitize-html](https://www.npmjs.com/package/sanitize-html) package on a need basis if HTML is turned on. This will
    whitelist only specific HTML tags needed for rich text markdown.

### Performance

Performance of tiptap explained in an example of having [long texts](https://tiptap.dev/examples/book).

### Dependencies

This component is dependent on the [`tiptap`](https://tiptap.dev/) third party library. As mentioned,
[tiptap](https://tiptap.dev/introduction#why-should-i-use-tiptap) is built on top of prosemirror, we might also use some of the prosemirror's

> library. For the currently supported features, we will include the following libraries that will be added to the packages.json

Reviewer 2: ```suggestion
library. For the currently supported features, we will include the following libraries that will be added to the package.json
```

    Below are a few scenarios to understand when this event will fire:

    1. This event will fired for every input in the content of the editor, including text inputs, text formatting changes, and text removals.
    2. This event will not fire when there are no changes made to the content of the editor. For example, all mouse events, selecting the texts, state
       changes, etc,

_CSS Classes and CSS Custom Properties that affect the component_

-   The minimum height of the component will be set to show at least one line in the text area along with a footer section. If the content exceeds
    the current height (set by the client), a vertical scrollbar will appear, allowing users to view the hidden content.
-   The minimum width of the component will be determined based on ensuring all buttons in the footer are visible. We have arrived to this decision
    to align with the minimal requirement for the initial release. However, adjustments will be made later based on the visual design for mobile view is complete.
-   The content in the text area will adjust its layout based on the width, potentially increasing the height of the editor. But in case of having

>     a long nested lists that is beyond the current width (set by the client), a horizontal scrollbar will appear within the text area to view the content.

Reviewer 4: Is this saying the control will always grow vertically to fit the content that is being edited without making a vertical scrollbar?

I think we should be using the native textarea as an example of the default behavior we should follow. If I understand correctly that the intention is the control will grow to fit the content I think that should be an optional configuration. Maybe a `<nimble-rich-text-editor auto-size>` or a `<nimble-rich-text-editor fit-to-content>` option

This applies to the viewer element as-well

       there are additional elements beyond this limit, they will be positioned below within the same footer-actions container.

_Note_: The positioning of these slot elements in the mobile view of the component has not yet been confirmed.

_Host Classes_

-   none

_Slotted Content/Slotted Classes_

-   none

_CSS Parts_

-   none

### `nimble-rich-text-viewer`

The `nimble-rich-text-viewer` is used for viewing rich text content when a markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### API

_Props/Attrs_

-   `markdownValue` - is an accessor used to get and set the markdown value.
    -   `getter` - this will merely return the markdown string that is set to the component.
    -   `setter` - this will parse the markdown string into a Node using
        [prosemirror-markdown parser](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/from_markdown.ts#L199) and convert to an HTML to
        render it in the component section.

_Methods_

-   none

_Events_

-   none

_CSS Classes and CSS Custom Properties that affect the component_

> -   The height of the component will always grow to fit the content.

Reviewer 4: Yea as mentioned above I think growing the height should be opt-in. It should have the same sizing behavior as the editor.

2. `footer` section consists of `nimble-toggle-button` to control each text formatting functionalities like bold, italic, etc, and a
   `footer-actions` slot element which is typically used to add action buttons to the right bottom of the component.

Example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>
    <nimble-button slot="footer-actions">Cancel</nimble-button>
    <nimble-button slot="footer-actions">Add Comment</nimble-button>
</nimble-rich-text-editor>
```

### API

_Props/Attrs_

-   `markdownValue` - is an accessor used to get and set the markdown value.
    -   `getter` - this will serialize the content by extracting the Node from the editor and convert it into a markdown string using
        [prosemirror-markdown serializer](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/to_markdown.ts#L30).
    -   `setter` - this will parse the markdown string into a Node and load it back into the editor using
        [prosemirror-markdown parser](https://github.com/ProseMirror/prosemirror-markdown/blob/master/src/from_markdown.ts#L199).

> -   `isEmpty` - is a read-only property that indicates whether the editor is empty or not. This will be achieved through Tiptap's

Reviewer 5: Not sure how critical it is to capture Angular implementation details in this spec, but I think it's at least worth mentioning how we would accomplish a bindable, read-only property in the Angular API. We would likely need to follow the pattern of the [`openChange` `Output` of the `NimbleBannerDirective`](https://github.com/ni/nimble/blob/1d5ab5f6f23e11e7b7c30600c196a04ea087c5af/angular-workspace/projects/ni/nimble-angular/src/directives/banner/nimble-banner.directive.ts#L25), though without the need for the `Change` suffix, as there is no two-way binding concern. Doing this requires handling an event from the underlying component in order to emit the event from the directive (as done [here](https://github.com/ni/nimble/blob/1d5ab5f6f23e11e7b7c30600c196a04ea087c5af/angular-workspace/projects/ni/nimble-angular/src/directives/banner/nimble-banner.directive.ts#L67) by the `NimbleBannerDirective`).

Author: For `isEmpty`, we have referred to the [read-only property `open`](https://github.com/ni/nimble/blob/1d5ab5f6f23e11e7b7c30600c196a04ea087c5af/packages/nimble-components/src/dialog/index.ts#L81) of the nimble dialog ([spec reference](https://github.com/ni/nimble/blob/1d5ab5f6f23e11e7b7c30600c196a04ea087c5af/packages/nimble-components/src/dialog/specs/README.md?plain=1#L63)) which returns true if the dialog is open. The Angular implementation of the `isEmpty` will be same as the [`open` attribute of the `NimbleDialogDirective`](https://github.com/ni/nimble/blob/1d5ab5f6f23e11e7b7c30600c196a04ea087c5af/angular-workspace/projects/ni/nimble-angular/src/directives/dialog/nimble-dialog.directive.ts#L51). The `isEmpty` getter method in the component simply returns the Tiptap's [isEmpty](https://tiptap.dev/api/editor#is-empty) value which will be true if the editor is empty.

With this approach, we can directly bind the `isEmpty` in the HTML of the Angular application. As an example, in the comments feature, we have the requirement to disable the `Comment` button when the editor is empty. So we can do something like the below code in the client to enable or disable the button.
```html
<nimble-rich-text-editor #editor>
    <nimble-button slot="footer-actions">Cancel</nimble-buttont>
    <nimble-button slot="footer-actions" [disabled]="editor.isEmpty">Comment</nimble-buttont>
</nimble-rich-text-editor>
```

So we thought of adding it as a read-only property rather than an event. We will update the spec to mention the Angular implementation for this API.

Reviewer 2: Resolving this thread since I think it's a sufficient level of detail for the spec. @atmgrifter00 mentioned offline that he still wanted to understand this a bit better, but it doesn't need to block initial submission (we can always update the spec if our plans change).

[Comments UI mockup](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045)

[Comments Feature](https://dev.azure.com/ni/DevCentral/_backlogs/backlog/ASW%20SystemLink%20Platform/Initiatives/?workitem=2205215)

### Non-goals

-   Blazor integration will be out of scope for this component as we have less or no experience in Blazor component development or other
    related technology. Blazor example app addition will also be not covered in this spec.

### Features

> Both the components provides support for the basic text formatting options as follows:

Reviewer: nit:
```suggestion
Both components provide support for the following basic text formatting options:
```

1. Bold
2. Italics
3. Numbered List
4. Bulleted List

> The `nimble-rich-text-editor` component will also offer APIs and interactive methods to format texts in the following ways:

Reviewer: nit:
```suggestion
The `nimble-rich-text-editor` component will also offer APIs and interactive methods to format text in the following ways:
```
